src/Caching/StackExchangeRedis/src/RedisOutputCacheStore.cs

```diff
@@ -0,0 +1,487 @@
+    private readonly RedisKey _valueKeyPrefix, _tagKeyPrefix, _tagMasterKey;
+    private readonly RedisKey[] _tagMasterKeyArray; // for use with Lua if needed (to avoid array allocs)
+    private readonly SemaphoreSlim _connectionLock = new SemaphoreSlim(initialCount: 1, maxCount: 1);
+
+    private bool _disposed;
+    private volatile IDatabase? _cache;
+    private long _lastConnectTicks = DateTimeOffset.UtcNow.Ticks;
+    private long _firstErrorTimeTicks;
+    private long _previousErrorTimeTicks;
+    private bool _useMultiExec, _use62Features;
+
+    internal bool GarbageCollectionEnabled { get; set; } = true;
+
+    // Never reconnect within 60 seconds of the last attempt to connect or reconnect.
+    private readonly TimeSpan ReconnectMinInterval = TimeSpan.FromSeconds(60);
+    // Only reconnect if errors have occurred for at least the last 30 seconds.
+    // This count resets if there are no errors for 30 seconds
+    private readonly TimeSpan ReconnectErrorThreshold = TimeSpan.FromSeconds(30);
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="RedisCache"/>.
+    /// </summary>
+    /// <param name="optionsAccessor">The configuration options.</param>
+    public RedisOutputCacheStore(IOptions<RedisCacheOptions> optionsAccessor) // TODO: OC-specific options?
+        : this(optionsAccessor, Logging.Abstractions.NullLoggerFactory.Instance.CreateLogger<RedisCache>())
+    {
+    }
+
+#if DEBUG
+    internal async ValueTask<string> GetConfigurationInfoAsync(CancellationToken cancellationToken = default)
+    {
+        await ConnectAsync(cancellationToken).ConfigureAwait(false);
+        return $"redis output-cache; MULTI/EXEC: {_useMultiExec}, v6.2+: {_use62Features}";
+    }
+#endif
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="RedisCache"/>.
+    /// </summary>
+    /// <param name="optionsAccessor">The configuration options.</param>
+    /// <param name="logger">The logger.</param>
+    internal RedisOutputCacheStore(IOptions<RedisCacheOptions> optionsAccessor, ILogger logger)
+    {
+        ArgumentNullThrowHelper.ThrowIfNull(optionsAccessor);
+        ArgumentNullThrowHelper.ThrowIfNull(logger);
+
+        _options = optionsAccessor.Value;
+        _logger = logger;
+
+        // This allows partitioning a single backend cache for use with multiple apps/services.
+
+        // SE.Redis allows efficient append of key-prefix scenarios, but we can help it
+        // avoid some work/allocations by forcing the key-prefix to be a byte[]; SE.Redis
+        // would do this itself anyway, using UTF8
+        _valueKeyPrefix = (RedisKey)Encoding.UTF8.GetBytes(_options.InstanceName + "__MSOCV_");
+        _tagKeyPrefix = (RedisKey)Encoding.UTF8.GetBytes(_options.InstanceName + "__MSOCT_");
+        _tagMasterKey = (RedisKey)Encoding.UTF8.GetBytes(_options.InstanceName + "__MSOCT");
+        _tagMasterKeyArray = new[] { _tagMasterKey };
+
+        _ = Task.Run(RunGarbageCollectionLoopAsync);
```

Comment:
Should this use the `Task.Factory.StartNew` with `TaskCreationOptions.LongRunning` in order to don't use a thread-pool thread for the long running (~ process lifetime) loop?

Reply:
applied